cpp/src/strings/replace/replace.cu

```diff
@@ -80,33 +101,436 @@ struct replace_fn {
+ * @return Number of target positions after filtering.
+ */
+size_type filter_false_target_positions(rmm::device_uvector<size_type>& target_positions,
+                                        device_span<int32_t const> d_offsets_span,
+                                        size_type target_size,
+                                        rmm::cuda_stream_view stream)
+{
+  // In-place remove any positions for target strings that crossed string boundaries.
+  auto d_target_positions = target_positions.data();
+  auto target_pos_end =
+    thrust::remove_if(rmm::exec_policy(stream),
+                      d_target_positions,
+                      d_target_positions + target_positions.size(),
+                      [d_offsets_span, target_size] __device__(size_type target_pos) -> bool {
+                        // find the end of the string containing the start of this target
+                        size_type const* offset_ptr = thrust::upper_bound(
+                          thrust::seq, d_offsets_span.begin(), d_offsets_span.end(), target_pos);
+                        return target_pos + target_size >= *offset_ptr;
+                      });
+  auto const target_count = cudf::distance(d_target_positions, target_pos_end);
+  if (target_count == 0) { return 0; }
+
+  // Filter out target positions that are the result of overlapping target matches.
+  return (target_count > 1)
+           ? filter_overlap_target_positions(d_target_positions, target_count, target_size, stream)
+           : target_count;
+}
+
+/**
+ * @brief Filter target positions beyond the maximum target replacements per row limit.
+ *
+ * This performs an in-place modification of the target positions to remove any target positions
+ * corresponding to targets that should not be replaced due to the maximum target replacement per
+ * row limit.
+ *
+ * @param[in,out] target_positions Target positions to filter in-place.
+ * @param[in]     target_count     Number of target positions.
+ * @param[in]     d_offsets_span   Memory range encompassing the string column offsets.
+ * @param[in]     max_repl_per_row Maximum target replacements per row limit.
+ * @param[in]     stream           CUDA stream to use for device operations.
+ * @return Number of target positions after filtering.
+ */
+size_type filter_maxrepl_target_positions(size_type* d_target_positions,
+                                          size_type target_count,
+                                          device_span<int32_t const> d_offsets_span,
+                                          size_type max_repl_per_row,
+                                          rmm::cuda_stream_view stream)
+{
+  auto pos_to_row_fn = [d_offsets_span] __device__(size_type target_pos) -> size_type {
+    auto upper_bound =
+      thrust::upper_bound(thrust::seq, d_offsets_span.begin(), d_offsets_span.end(), target_pos);
+    return upper_bound - d_offsets_span.begin();
```

Comment:
```suggestion
    return cudf::distance(d_offsets_span.begin(), upper_bound);
```

Reply:
Does cudf::distance have a `__device__` implementation?  This change doesn't compile as-is.

Reply:
Ah. I thought it was `constexpr`. You should be able to use `thrust::distance()` instead.